pkg/blockbuilder/blockbuilder.go

```diff
@@ -239,3 +240,308 @@ func (b *BlockBuilder) getLagForPartition(ctx context.Context, partition int32)
+			"last_block_end", time.UnixMilli(startState.LastBlockEnd), "curr_block_end", time.UnixMilli(state.LastBlockEnd),
+			"last_seen_till", startState.SeenTillOffset, "curr_seen_till", state.SeenTillOffset,
+			"num_blocks", numBlocks, "compact_and_upload_dur", compactionDur)
+	}(time.Now(), *state)
+
+	builder := NewTSDBBuilder(b.logger, b.cfg.DataDir, b.cfg.BlocksStorage, b.limits)
+	defer builder.Close()
+
+	level.Info(b.logger).Log("msg", "start consuming", "partition", state.Partition, "offset", state.Commit.At, "lag", state.Lag, "cycle_end", cycleEnd)
+
+	// TopicPartition to resume consuming on this iteration.
+	tp := map[string][]int32{b.cfg.Kafka.Topic: {state.Partition}}
+	b.kafkaClient.ResumeFetchPartitions(tp)
+	defer b.kafkaClient.PauseFetchPartitions(tp)
+
+	// We always rewind the partition's offset to the commit offset. This is so the cycle started exactly at the commit point,
+	// and not at what was (potentially over-) consumed previously.
+	b.seekPartition(state.Partition, state.Commit.At)
+
+	var (
+		firstRec  *kgo.Record
+		lastRec   *kgo.Record
+		commitRec *kgo.Record
+	)
+
+consumerLoop:
+	for remaining := state.Lag; remaining > 0; {
+		if err := context.Cause(ctx); err != nil {
+			return err
+		}
+		fetches := b.kafkaClient.PollFetches(ctx)
+		if fetches.IsClientClosed() {
+			level.Warn(b.logger).Log("msg", "client closed when fetching records")
+			return nil
+		}
+		fetches.EachError(func(_ string, _ int32, err error) {
+			if !errors.Is(err, context.Canceled) {
+				level.Error(b.logger).Log("msg", "failed to fetch records", "partition", state.Partition, "err", err)
+				b.metrics.fetchErrors.Inc()
+			}
+		})
+
+		numRecs := fetches.NumRecords()
+		remaining -= int64(numRecs)
+		b.metrics.fetchRecordsTotal.Add(float64(numRecs))
+
+		recIter := fetches.RecordIter()
+		for !recIter.Done() {
+			rec := recIter.Next()
+
+			if firstRec == nil {
+				firstRec = rec
+			}
+
+			level.Debug(b.logger).Log("msg", "consumed record", "offset", rec.Offset, "rec_ts", rec.Timestamp, "last_bmax", state.LastBlockEnd, "bmax", blockEnd)
+
+			// Stop consuming after we reached the cycleEnd marker.
+			// NOTE: the timestamp of the record is when the record was produced relative to distributor's time.
+			if rec.Timestamp.After(cycleEnd) {
+				break consumerLoop
+			}
+
+			recProcessedBefore := rec.Offset <= state.SeenTillOffset
+			allSamplesProcessed, err := builder.Process(ctx, rec, state.LastBlockEnd, blockEnd, recProcessedBefore)
+			if err != nil {
+				// All "non-terminal" errors are handled by the TSDBBuilder.
+				return fmt.Errorf("process record in partition %d at offset %d: %w", rec.Partition, rec.Offset, err)
+			}
+			if !allSamplesProcessed {
+				if lastRec == nil {
+					// The first record was not fully processed, meaning the record before this is the commit point.
+					// We have to re-commit this one to update the seenTillOffset; this is so the next cycle handled partially processed record properly.
```

Comment:
I've rephrased this comment. The point here is that we re-commit the same offset to update its metadata, with the latest value of `oldestSeenOffset`. Otherwise, committing what has been already committed on the previous cycle is redundant.